### PR TITLE
[#15] Add cardano-chain Delegation and other modules

### PR DESCRIPTION
**Pull request by ruhatch**
Add delegation modules from `cardano-sl`

### CODE REVIEW

File: src/Cardano/Chain/ProtocolConstants.hs [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+       , pcEpochSlots
+       , kEpochSlots
+       ) where
+
+import           Cardano.Prelude
+
+import           Control.Monad.Except (MonadError)
+import qualified Data.Aeson as Aeson (FromJSON (..), ToJSON (..))
+import           Text.JSON.Canonical (FromJSON (..), ToJSON (..))
```
Comment (redxaxder) [inline src/Cardano/Chain/ProtocolConstants.hs:34]: So there are two sets of `FromJSON` and `ToJSON`. 
The type in this module has an instance for each, and in `Cardano.Chain.Delegation.HeavyDlgIndex` we have:

 - `HeavyDlgIndex`, with an aeson instance, but no canonical-json instance
- `ProxySKHeavy`, with the opposite

What's the deal with these?

Comment (ruhatch) [reply src/Cardano/Chain/ProtocolConstants.hs:34]: The `Aeson` instances were used for logging I believe, but we might remove them all together. The `canonical-json` instances are used for decoding the `GenesisData`, a bunch of initial parameters stored in the genesis block.

Comment (redxaxder) [reply src/Cardano/Chain/ProtocolConstants.hs:34]: Is `Buildable` also used for logging?

Comment (ruhatch) [reply src/Cardano/Chain/ProtocolConstants.hs:34]: Yeah - some decisions either have been or need to made about what to keep

**Review by mdimjasevic — APPROVED** [review]

**Pull request merged**